lib/src/cache/cache.dart

```diff
@@ -0,0 +1,245 @@
+  /// Identifies a single instance of a [TValue]
+  final TIdentifier id;
+
+  /// The current value stored in the [Cache] for the attributed [TIdentifier].
+  final TValue value;
+
+  CacheContext(this.id, this.value);
+}
+
+/// Maintians a reference to a given any [TValue] attributed to a [TIdentifier].
+///
+/// References are retained for the lifecycle of the instance of the [Cache],
+/// unless explicitly removed.
+class Cache<TIdentifier, TValue> extends Object with Disposable {
+  /// The backing store for values in the [Cache].
+  Map<TIdentifier, Future<TValue>> _cache = <TIdentifier, Future<TValue>>{};
+
+  // ignore: close_sinks
+  StreamController<CacheContext<TIdentifier, TValue>> _didUpdateController =
+      new StreamController<CacheContext<TIdentifier, TValue>>.broadcast();
+
+  // ignore: close_sinks
+  StreamController<CacheContext<TIdentifier, TValue>> _didRemoveController =
+      new StreamController<CacheContext<TIdentifier, TValue>>.broadcast();
+
+  /// A collection of pending removals for a given [TIdentifier].
+  Map<TIdentifier, Completer<Null>> _removalCompleters =
+      <TIdentifier, Completer<Null>>{};
+
+  Cache() {
+    [_didRemoveController, _didUpdateController]
+        .forEach(manageStreamController);
+  }
+
+  /// The collection of removals from the [Cache].
+  Stream<CacheContext<TIdentifier, TValue>> get didRemove =>
+      _didRemoveController.stream;
+
+  /// The collection of changes to the [Cache].
+  ///
+  /// Values are emitted when the cached value is changes.
+  Stream<CacheContext<TIdentifier, TValue>> get didUpdate =>
+      _didUpdateController.stream;
+
+  /// Returns a value from the cache for a given [TIdentifier] or adds the
+  /// [TValue] returned by the value factory when the value does not exist in
+  /// the cache.
+  ///
+  /// All calls to [get] await a call to the [onGet] lifecycle method before
+  /// returning the cached value. If the [Cache] does not contain an instance
+  /// for the given [TIdentifier], the given [valueFactory] is called and an
+  /// [CacheContext] event is emitted on the [didUpdate] stream. A call to [get]
+  /// that returns a cached value does not emit this event as the [Cache] has
+  /// not updated.
+  ///
+  /// Calls to [get] are evaluated syncronously and will return the future value
+  /// established by the first call to [get]. This allows calls to get to
+  /// return the same value regardless of completion of asyncronous calls to
+  /// the [onGet] lifecycle methods or [valueFactory].
+  ///
+  /// ```
+  /// var a = cache.get('id', _superLongCall);
+  /// var b = cache.get('id', _superLongCall);
+  /// var c = cache.get('id', _superLongCall);
+  ///
+  /// // All of the values in this list are the same instance returned by the
+  /// // first call to `_superLongCall`.
+  /// var values = Future.wait([a, b, c]);
+  /// ```
+  ///
+  /// If the [Cache] [isDisposedOrDisposing] then a [StateError] is thrown.
```

Comment:
I went with the reference as it allows consumers to jump to the definition which is explicit in what that reference means.